### PR TITLE
actor codes

### PR DESCRIPTION
**Pull request by turuslan**
- `ActorCodeCid` strong type
- remove old "v*/codes.hpp" files
- `FC_OPERATOR_NOT_EQUAL` macro to complement `operator==`
- fix market actor test

### CODE REVIEW

File: core/primitives/cid/cid.cpp [hunk 1]
```diff
@@ -150,6 +156,15 @@ namespace fc {
     return {};
   }
 
+  boost::optional<ActorCodeCid> asActorCode(const CID &cid) {
+    if (auto id{asIdentity(cid)}) {
+      if (cid.content_type == CID::Multicodec::RAW) {
+        return ActorCodeCid{common::span::bytestr(*id)};
+      }
+    }
+    return {};
```
Comment (ortyomka) [inline core/primitives/cid/cid.cpp:165]:
```suggestion
    return boost::none;
```

File: core/vm/actor/code.hpp [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+ */
+
+#pragma once
+
+#include <string_view>
+
+namespace fc {
+  struct ActorCodeCid : std::string_view {};
+  constexpr bool operator==(const ActorCodeCid &l, const ActorCodeCid &r) {
+    return (const std::string_view &)l == (const std::string_view &)r;
```
Comment (ortyomka) [inline core/vm/actor/code.hpp:13]: With `static_cast`

File: core/primitives/cid/cid.cpp [hunk 3]
```diff
@@ -150,6 +156,15 @@ namespace fc {
     return {};
   }
 
+  boost::optional<ActorCodeCid> asActorCode(const CID &cid) {
+    if (auto id{asIdentity(cid)}) {
+      if (cid.content_type == CID::Multicodec::RAW) {
```
Comment (Alexey-N-Chernyshov) [inline core/primitives/cid/cid.cpp:161]: Did you forget to check cid version?

File: core/vm/actor/codes.hpp [hunk 4]
```diff
@@ -58,3 +58,23 @@ namespace fc::vm::actor::code {
...
+    constexpr auto kInitCodeId{code::init##V};                 \
+    constexpr auto kPaymentChannelCodeId{code::paych##V};      \
+    constexpr auto kRewardActorCodeId{code::reward##V};        \
+    constexpr auto kSystemActorCodeId{code::system##V};        \
+    constexpr auto kVerifiedRegistryCodeId{code::verifreg##V}; \
+  }
+VM_ACTOR_BUILTIN_V_CODE(0)
+VM_ACTOR_BUILTIN_V_CODE(2)
+VM_ACTOR_BUILTIN_V_CODE(3)
+VM_ACTOR_BUILTIN_V_CODE(4)
```
Comment (Alexey-N-Chernyshov) [inline core/vm/actor/codes.hpp:79]: Is it time to add v5 now?

**Review by Alexey-N-Chernyshov — APPROVED** [review]

File: core/vm/actor/codes.hpp (hunk 4 above)
Comment (turuslan) [reply core/vm/actor/codes.hpp:79]: Yes, in other pr (with actor/network version and others)

File: core/primitives/cid/cid.cpp (hunk 3 above)
Comment (turuslan) [reply core/primitives/cid/cid.cpp:161]: Thanks, will check

File: core/common/cmp.hpp [hunk 5]
```diff
@@ -5,13 +5,19 @@
 
 #pragma once
 
+#define FC_OPERATOR_NOT_EQUAL_2(L, R)              \
```
Comment (wer1st) [inline core/common/cmp.hpp:8]: What is a problem is resolved by this solution?
I don't think that operator != is too big or hard to hide it under the macros

**Review by ortyomka — APPROVED** [review]

File: core/common/cmp.hpp (hunk 5 above)
Comment (Alexey-N-Chernyshov) [reply core/common/cmp.hpp:8]: similar to 'boost::equality_comparable'

**Pull request merged**